File: umwia2004-java/src/main/java/com/umwia2004/solution/lab1_2/compiled/Process.java
package com.umwia2004.solution.lab1_2.compiled;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.time.Duration;
import java.time.LocalDateTime;

@Getter
@RequiredArgsConstructor
public class Process {
    private final String processId;
    private final LocalDateTime arrivalTime;
    private final Duration burstTime;

    private LocalDateTime startTime;
    private Duration turnaroundTime;
    private Duration waitingTime;

    public void executeAt(LocalDateTime startTime) {
        this.startTime = startTime;
        this.waitingTime = Duration.between(arrivalTime, startTime);
        this.turnaroundTime = waitingTime.plus(burstTime);
    }
}

File: umwia2004-rust/src/lab1/process.rs
#[derive(Clone)]
pub struct Process {
    pub process_id: String,
    pub arrival_time: f64,
    pub burst_time: f64,
    pub turnaround_time: f64,
    pub waiting_time: f64,
}

impl Process {
    pub fn new(process_id: &str, arrival_time: f64, burst_time: f64) -> Self {
        Self {
            process_id: process_id.to_string(),
            arrival_time,
            burst_time,
            turnaround_time: 0.0,
            waiting_time: 0.0,
        }
    }

    pub fn update_times(&mut self, start_time: f64) {
        self.waiting_time = (start_time - self.arrival_time).max(0.0);
        self.turnaround_time = self.waiting_time + self.burst_time;
    }
}
